configure.ac

```diff
@@ -445,6 +445,7 @@ AC_ARG_ENABLE(fast,
         alwaysinline - Force compiler to always inline performance critical routines
         sse      - Use SSE2-based copy routines
         avx      - Use AVX256-based copy routines
+        avx512   - Use AVX512-based copy routines
```

Comment:
enable avx or avx512 are too ambiguous -- I know, not your fault ;) The instruction set can be used in multiple features and the `--enable-` options should be per-feature. How about something like `--enable-ch4-shm-nontemporal-copy` (shorter would be nice) then inside configure we check for either avx512 or avx?

Reply:
Yes, there options are not very straight forward. There are two set of functionality that MPICH uses: 1) non-temporal store intrinsics with AVX256 or SSE2 (used in POSIX) and 2) regular mem copy with AVX256 or AVX512 (used in mpl/gpu/mpl_gpu_ze.c).

Thinking about the problem I run into before this PR. I think users would expect `--enable-fast` to be portable, some optimization options may have no effect, code built should behave as if those options were not given. I think the biggest problem is the `avx`, `sse` or `avx512` removes the portability of the MPICH build. It checks if AVX can build and run on the building machine, not the target machine. So if I build on a intel machine, configure will detect AVX512 supported and build with it. If this build is then run on a AMD machine, it will fail because the lack of AVX512 instructions. While is not common that login nodes and compute nodes to have drastically different architecture.

Maybe something like `--enable-sse`, `--enable-avx` and `--enable-avx512` will be a better solution. We can note that these options enable the use of corresponding instructions, which limits the portability of the build. `--enable-sse` should be largely portable among x86, `--enable-avx` will be OK for most Intel CPUs and AMD CPUs starting Zen3, and `--enable-avx512` will be the most limited in portability as only a few Intel CPUs and AMD Zen5 (I think) support it.

More feature specific options like `--enable-ch4-shm-nontemporal-copy` or `--enable-ch4-shm-nt-copy` will be useful for a case where we enable AVX256 but do not want to use non-temporal copy. It will not work the other way, where we only want non-temporal copy feature and no other AVX256 features.

So, I think maybe we should just add `--enable-sse`, `--enable-avx` and `--enable-avx512` to turn on/off the usage of these instruction sets. The specific checkings on individual intrinsic maybe redundant. For example, if a x86 machine has AVX256 support, it should always have support for _mm256_stream_si256. It is probably only useful is some older compiler only implemented partial support of AVX256.

I will redo the PR.

Reply:
Or, we can have `--enable-ch4-shm-nt-copy` and `--enable-mpl-avx-move` to check if the corresponding intrinsics can be built without adding `-mavx` or `-mavx512f` to `CFLAGS`, and shift the responsibility of adding these flags to the user who build MPICH. And we can warn user about missing if those feature are enabled but CFLAGS are missing.

Not the most user-friendly choice IMO.